### ISSUE TITLE
Getting name updates to persist in the database

### ISSUE THREAD

**Issue by Ceisen9** [ga-wdi-exercises/project4#159]
I am having trouble getting my updates to persist in my database in my MEAN app. Right now all updates, deletes, etc are being handled by the show controller. The delete function does persist in the database, but the update function does not, although their code is functionally similar. The update function will show up on the page, but is forgotten upon reload. Here is my code for my show controller in my angular module file: 

```
function showShowCtrl(Show, $stateParams, $window){
    var vm = this;
    Show.find("name", $stateParams.name, function(show){
      vm.show = show;
    });
    vm.update = function(){
      Show.update({name: vm.show.name}, {show: vm.show}, function() {
        console.log("updated");
      });
    }
    vm.delete = function(){
      Show.remove({name: vm.show.name}, function(){
        $window.location.replace("/shows");
      });
    }
    vm.addEpisode = function(){
      if(vm.show.episodes.includes(vm.newEpisode)){
        console.log("nope, duplicate")
      } else{
        vm.show.episodes.push(vm.newEpisode);
        vm.newEpisode = "";
        vm.update();
      }
      vm.removeEpisode = function($index){
        vm.show.episodes.splice($index, 1);
        vm.update();
      }
    }
  }
```

And here are my routes defined on the back end in my index.js file:

```
app.delete("/api/shows/:name", function(req, res){
  Show.findOneAndRemove({name: req.params.name}).then(function(){
    res.json({success: true});
  });
});

app.put("/api/shows/:name", function(req, res){
  Show.findOneAndUpdate({name: req.params.name}, req.body.show, {new: true}).then(function(show){
    res.json(show);
  });
});
```

Please let me know how I can fix this issue!

**Comment by RobertAKARobin** [comment]
Try console logging the response Angular gets from Express:

``` js
Show.update({name: vm.show.name}, {show: vm.show}, function(response) {
    console.log(response);
});
```

What's it look like?

**Comment by Ceisen9** [comment]
```
ServerResponse {
  domain: null,
  _events: { finish: [Function: resOnFinish] },
  _eventsCount: 1,
  _maxListeners: undefined,
  output: [],
  outputEncodings: [],
  outputCallbacks: [],
  outputSize: 0,
  writable: true,
  _last: false,
  chunkedEncoding: false,
  shouldKeepAlive: true,
  useChunkedEncodingByDefault: true,
  sendDate: true,
  _removedHeader: { 'content-length': false },
  _contentLength: 4,
  _hasBody: true,
  _trailer: '',
  finished: true,
  _headerSent: true,
  socket: 
   Socket {
     _connecting: false,
     _hadError: false,
     _handle: 
      TCP {
        _externalStream: {},
        fd: 28,
        reading: true,
        owner: [Circular],
        onread: [Function: onread],
        onconnection: null,
        writeQueueSize: 0 },
     _parent: null,
     _host: null,
     _readableState: 
      ReadableState {
        objectMode: false,
        highWaterMark: 16384,
        buffer: [],
        length: 0,
        pipes: null,
        pipesCount: 0,
        flowing: true,
        ended: false,
        endEmitted: false,
        reading: true,
        sync: false,
        needReadable: true,
        emittedReadable: false,
        readableListening: false,
        resumeScheduled: false,
        defaultEncoding: 'utf8',
        ranOut: false,
        awaitDrain: 0,
        readingMore: false,
        decoder: null,
        encoding: null },
     readable: true,
     domain: null,
     _events: 
      { end: [Object],
        finish: [Function: onSocketFinish],
        _socketEnd: [Function: onSocketEnd],
        drain: [Object],
        timeout: [Function],
        error: [Object],
        close: [Object],
        data: [Function: socketOnData],
        resume: [Function: onSocketResume],
        pause: [Function: onSocketPause] },
     _eventsCount: 10,
     _maxListeners: undefined,
     _writableState: 
      WritableState {
        objectMode: false,
        highWaterMark: 16384,
        needDrain: false,
        ending: false,
        ended: false,
        finished: false,
        decodeStrings: false,
        defaultEncoding: 'utf8',
        length: 0,
        writing: false,
        corked: 0,
        sync: false,
        bufferProcessing: false,
        onwrite: [Function],
        writecb: null,
        writelen: 0,
        bufferedRequest: null,
        lastBufferedRequest: null,
        pendingcb: 3,
        prefinished: false,
        errorEmitted: false,
        bufferedRequestCount: 0,
        corkedRequestsFree: [Object] },
     writable: true,
     allowHalfOpen: true,
     destroyed: false,
     bytesRead: 0,
     _bytesDispatched: 4639,
     _sockname: null,
     _pendingData: null,
     _pendingEncoding: '',
     server: 
      Server {
        domain: null,
        _events: [Object],
        _eventsCount: 3,
        _maxListeners: undefined,
        _connections: 2,
        _handle: [Object],
        _usingSlaves: false,
        _slaves: [],
        _unref: false,
        allowHalfOpen: true,
        pauseOnConnect: false,
        httpAllowHalfOpen: false,
        timeout: 120000,
        _pendingResponseData: 0,
        _connectionKey: '6::::3001' },
     _server: 
      Server {
        domain: null,
        _events: [Object],
        _eventsCount: 3,
        _maxListeners: undefined,
        _connections: 2,
        _handle: [Object],
        _usingSlaves: false,
        _slaves: [],
        _unref: false,
        allowHalfOpen: true,
        pauseOnConnect: false,
        httpAllowHalfOpen: false,
        timeout: 120000,
        _pendingResponseData: 0,
        _connectionKey: '6::::3001' },
     _idleTimeout: 120000,
     _idleNext: 
      { [Function: utcDate]
        _onTimeout: [Function],
        _idleTimeout: 729,
        _idleNext: [Object],
        _idlePrev: [Circular],
        _idleStart: 14736,
        _called: true },
     _idlePrev: { _idleNext: [Circular], _idlePrev: [Object] },
     _idleStart: 14736,
     parser: 
      HTTPParser {
        '0': [Function: parserOnHeaders],
        '1': [Function: parserOnHeadersComplete],
        '2': [Function: parserOnBody],
        '3': [Function: parserOnMessageComplete],
        '4': [Function: onParserExecute],
        _headers: [],
        _url: '',
        _consumed: true,
        socket: [Circular],
        incoming: [Object],
        outgoing: null,
        maxHeaderPairs: 2000,
        onIncoming: [Function: parserOnIncoming] },
     on: [Function: socketOnWrap],
     _paused: false,
     read: [Function],
     _consuming: true,
     _httpMessage: [Circular] },
  connection: 
   Socket {
     _connecting: false,
     _hadError: false,
     _handle: 
      TCP {
        _externalStream: {},
        fd: 28,
        reading: true,
        owner: [Circular],
        onread: [Function: onread],
        onconnection: null,
        writeQueueSize: 0 },
     _parent: null,
     _host: null,
     _readableState: 
      ReadableState {
        objectMode: false,
        highWaterMark: 16384,
        buffer: [],
        length: 0,
        pipes: null,
        pipesCount: 0,
        flowing: true,
        ended: false,
        endEmitted: false,
        reading: true,
        sync: false,
        needReadable: true,
        emittedReadable: false,
        readableListening: false,
        resumeScheduled: false,
        defaultEncoding: 'utf8',
        ranOut: false,
        awaitDrain: 0,
        readingMore: false,
        decoder: null,
        encoding: null },
     readable: true,
     domain: null,
     _events: 
      { end: [Object],
        finish: [Function: onSocketFinish],
        _socketEnd: [Function: onSocketEnd],
        drain: [Object],
        timeout: [Function],
        error: [Object],
        close: [Object],
        data: [Function: socketOnData],
        resume: [Function: onSocketResume],
        pause: [Function: onSocketPause] },
     _eventsCount: 10,
     _maxListeners: undefined,
     _writableState: 
      WritableState {
        objectMode: false,
        highWaterMark: 16384,
        needDrain: false,
        ending: false,
        ended: false,
        finished: false,
        decodeStrings: false,
        defaultEncoding: 'utf8',
        length: 0,
        writing: false,
        corked: 0,
        sync: false,
        bufferProcessing: false,
        onwrite: [Function],
        writecb: null,
        writelen: 0,
        bufferedRequest: null,
        lastBufferedRequest: null,
        pendingcb: 3,
        prefinished: false,
        errorEmitted: false,
        bufferedRequestCount: 0,
        corkedRequestsFree: [Object] },
     writable: true,
     allowHalfOpen: true,
     destroyed: false,
     bytesRead: 0,
     _bytesDispatched: 4639,
     _sockname: null,
     _pendingData: null,
     _pendingEncoding: '',
     server: 
      Server {
        domain: null,
        _events: [Object],
        _eventsCount: 3,
        _maxListeners: undefined,
        _connections: 2,
        _handle: [Object],
        _usingSlaves: false,
        _slaves: [],
        _unref: false,
        allowHalfOpen: true,
        pauseOnConnect: false,
        httpAllowHalfOpen: false,
        timeout: 120000,
        _pendingResponseData: 0,
        _connectionKey: '6::::3001' },
     _server: 
      Server {
        domain: null,
        _events: [Object],
        _eventsCount: 3,
        _maxListeners: undefined,
        _connections: 2,
        _handle: [Object],
        _usingSlaves: false,
        _slaves: [],
        _unref: false,
        allowHalfOpen: true,
        pauseOnConnect: false,
        httpAllowHalfOpen: false,
        timeout: 120000,
        _pendingResponseData: 0,
        _connectionKey: '6::::3001' },
     _idleTimeout: 120000,
     _idleNext: 
      { [Function: utcDate]
        _onTimeout: [Function],
        _idleTimeout: 729,
        _idleNext: [Object],
        _idlePrev: [Circular],
        _idleStart: 14736,
        _called: true },
     _idlePrev: { _idleNext: [Circular], _idlePrev: [Object] },
     _idleStart: 14736,
     parser: 
      HTTPParser {
        '0': [Function: parserOnHeaders],
        '1': [Function: parserOnHeadersComplete],
        '2': [Function: parserOnBody],
        '3': [Function: parserOnMessageComplete],
        '4': [Function: onParserExecute],
        _headers: [],
        _url: '',
        _consumed: true,
        socket: [Circular],
        incoming: [Object],
        outgoing: null,
        maxHeaderPairs: 2000,
        onIncoming: [Function: parserOnIncoming] },
     on: [Function: socketOnWrap],
     _paused: false,
     read: [Function],
     _consuming: true,
     _httpMessage: [Circular] },
  _header: 'HTTP/1.1 200 OK\r\nX-Powered-By: Express\r\nContent-Type: application/json; charset=utf-8\r\nContent-Length: 4\r\nETag: W/"4-N6YlnMDB2uKZp4Zkid/wvQ"\r\nDate: Mon, 25 Apr 2016 14:59:44 GMT\r\nConnection: keep-alive\r\n\r\n',
  _headers: 
   { 'x-powered-by': 'Express',
     'content-type': 'application/json; charset=utf-8',
     'content-length': '4',
     etag: 'W/"4-N6YlnMDB2uKZp4Zkid/wvQ"' },
  _headerNames: 
   { 'x-powered-by': 'X-Powered-By',
     'content-type': 'Content-Type',
     'content-length': 'Content-Length',
     etag: 'ETag' },
  _onPendingData: [Function: updateOutgoingData],
  req: 
   IncomingMessage {
     _readableState: 
      ReadableState {
        objectMode: false,
        highWaterMark: 16384,
        buffer: [],
        length: 0,
        pipes: null,
        pipesCount: 0,
        flowing: true,
        ended: true,
        endEmitted: true,
        reading: false,
        sync: true,
        needReadable: false,
        emittedReadable: false,
        readableListening: false,
        resumeScheduled: false,
        defaultEncoding: 'utf8',
        ranOut: false,
        awaitDrain: 0,
        readingMore: false,
        decoder: null,
        encoding: null },
     readable: false,
     domain: null,
     _events: {},
     _eventsCount: 0,
     _maxListeners: undefined,
     socket: 
      Socket {
        _connecting: false,
        _hadError: false,
        _handle: [Object],
        _parent: null,
        _host: null,
        _readableState: [Object],
        readable: true,
        domain: null,
        _events: [Object],
        _eventsCount: 10,
        _maxListeners: undefined,
        _writableState: [Object],
        writable: true,
        allowHalfOpen: true,
        destroyed: false,
        bytesRead: 0,
        _bytesDispatched: 4639,
        _sockname: null,
        _pendingData: null,
        _pendingEncoding: '',
        server: [Object],
        _server: [Object],
        _idleTimeout: 120000,
        _idleNext: [Object],
        _idlePrev: [Object],
        _idleStart: 14736,
        parser: [Object],
        on: [Function: socketOnWrap],
        _paused: false,
        read: [Function],
        _consuming: true,
        _httpMessage: [Circular] },
     connection: 
      Socket {
        _connecting: false,
        _hadError: false,
        _handle: [Object],
        _parent: null,
        _host: null,
        _readableState: [Object],
        readable: true,
        domain: null,
        _events: [Object],
        _eventsCount: 10,
        _maxListeners: undefined,
        _writableState: [Object],
        writable: true,
        allowHalfOpen: true,
        destroyed: false,
        bytesRead: 0,
        _bytesDispatched: 4639,
        _sockname: null,
        _pendingData: null,
        _pendingEncoding: '',
        server: [Object],
        _server: [Object],
        _idleTimeout: 120000,
        _idleNext: [Object],
        _idlePrev: [Object],
        _idleStart: 14736,
        parser: [Object],
        on: [Function: socketOnWrap],
        _paused: false,
        read: [Function],
        _consuming: true,
        _httpMessage: [Circular] },
     httpVersionMajor: 1,
     httpVersionMinor: 1,
     httpVersion: '1.1',
     complete: true,
     headers: 
      { host: 'localhost:3001',
        connection: 'keep-alive',
        'content-length': '131',
        pragma: 'no-cache',
        'cache-control': 'no-cache',
        accept: 'application/json, text/plain, */*',
        origin: 'http://localhost:3001',
        'user-agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/49.0.2623.112 Safari/537.36',
        'content-type': 'application/json;charset=UTF-8',
        referer: 'http://localhost:3001/shows/Survivor',
        'accept-encoding': 'gzip, deflate, sdch',
        'accept-language': 'en-US,en;q=0.8' },
     rawHeaders: 
      [ 'Host',
        'localhost:3001',
        'Connection',
        'keep-alive',
        'Content-Length',
        '131',
        'Pragma',
        'no-cache',
        'Cache-Control',
        'no-cache',
        'Accept',
        'application/json, text/plain, */*',
        'Origin',
        'http://localhost:3001',
        'User-Agent',
        'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/49.0.2623.112 Safari/537.36',
        'Content-Type',
        'application/json;charset=UTF-8',
        'Referer',
        'http://localhost:3001/shows/Survivor',
        'Accept-Encoding',
        'gzip, deflate, sdch',
        'Accept-Language',
        'en-US,en;q=0.8' ],
     trailers: {},
     rawTrailers: [],
     upgrade: false,
     url: '/api/shows/Survivortown',
     method: 'PUT',
     statusCode: null,
     statusMessage: null,
     client: 
      Socket {
        _connecting: false,
        _hadError: false,
        _handle: [Object],
        _parent: null,
        _host: null,
        _readableState: [Object],
        readable: true,
        domain: null,
        _events: [Object],
        _eventsCount: 10,
        _maxListeners: undefined,
        _writableState: [Object],
        writable: true,
        allowHalfOpen: true,
        destroyed: false,
        bytesRead: 0,
        _bytesDispatched: 4639,
        _sockname: null,
        _pendingData: null,
        _pendingEncoding: '',
        server: [Object],
        _server: [Object],
        _idleTimeout: 120000,
        _idleNext: [Object],
        _idlePrev: [Object],
        _idleStart: 14736,
        parser: [Object],
        on: [Function: socketOnWrap],
        _paused: false,
        read: [Function],
        _consuming: true,
        _httpMessage: [Circular] },
     _consuming: true,
     _dumped: false,
     next: [Function: next],
     baseUrl: '',
     originalUrl: '/api/shows/Survivortown',
     _parsedUrl: 
      Url {
        protocol: null,
        slashes: null,
        auth: null,
        host: null,
        port: null,
        hostname: null,
        hash: null,
        search: null,
        query: null,
        pathname: '/api/shows/Survivortown',
        path: '/api/shows/Survivortown',
        href: '/api/shows/Survivortown',
        _raw: '/api/shows/Survivortown' },
     params: { name: 'Survivortown' },
     query: {},
     res: [Circular],
     body: { show: [Object] },
     _body: true,
     length: undefined,
     read: [Function],
     route: Route { path: '/api/shows/:name', stack: [Object], methods: [Object] } },
  locals: {},
  statusMessage: 'OK',
  statusCode: 200 }
```

**Comment by Ceisen9** [comment]
I changed the tv shows name from Survivor to Survivortown in that instance

**Comment by RobertAKARobin** [comment]
Gahhh! Wrong place. Console log in your Angular controller.

**Comment by Ceisen9** [comment]
Not sure how I can show you the angular response from the console without it looking like gibberish, but I get an object with the name correctly replaced. However, it has the information that "object value at left was snapshotted when logged, value below was evaluated just now"

**Comment by Ceisen9** [comment]
Just got it to work by changing .findOneAndUpdate to .update... Will this lead to any problems down the line?

**Comment by Ceisen9** [comment]
Wait, I spoke too soon. It's still not persisting in the db

**Comment by RobertAKARobin** [comment]
I think you're getting confused between Angular methods and Mongoose methods. Can you push your code to Github and link to it here?

**Comment by Ceisen9** [comment]
https://github.com/Ceisen9/show-stopper

**Comment by RobertAKARobin** [comment]
Found it! The problem is in this line:

``` js
Show.update({name: vm.show.name}, {show: vm.show}
```

`vm.show.name` changes when you type something into the form, e.g. when I change "Chappelle's Show" to "Clam chowder" Angular changes `vm.show.name` to "Clam chowder".

That then gets passed to Express, which tries to `findOneAndUpdate` a show called "Clam chowder", which doesn't exist yet. You want to tell the database to find "Chappelle's Show" and _then_ update it to "Clam chowder".

So in `Show.update` the first argument should be the show's **old** name, and the second argument should be the **new** data.

TLDR: Change it to this:

``` js
Show.update({name: $stateParams.name}, {show: vm.show}
```

**Comment by Ceisen9** [comment]
Haha you and your proxy names... Thanks Robin!
